### ISSUE TITLE
Benchmark and optimize on large datasets

### ISSUE THREAD

**Issue by rossant** [kwikteam/phy#129]
* [x] Choose and download locally one of Nick's huge datasets.
* [x] Optimize the cluster store (generation time and loading time)
    * Initialize all arrays in the cluster store
    * Loop over all spikes
    * Copy over the features and masks in the cluster store
* [x] Benchmark

**Comment by rossant** [comment]
Some work has been done in #160 concerning benchmarking and optimization of the cluster store generation (happens only once, when a dataset is first loaded into phy).

Recall that the cluster store stores one HDF5 file per cluster. This file contains the features and masks of that cluster in dense format (for now). The files are currently saved to `~/.phy/cluster_store/my_dataset/`. The cluster store is generated when opening a dataset for the first time on a given computer. Re-opening the dataset on the same computer allows to reuse the existing cluster store. There is some partial consistency check to be sure that the KWIK file and the cluster store haven't diverged, otherwise the cluster store has to be re-generated.

The current implementation uses very little RAM but is very slow. The HDF5 cluster files are first initialized and contain empty arrays of the right size. Then, there is a loop over spikes. For every spike, the features and masks are loaded from the KWX file and copied to the corresponding HDF5 cluster file.

On small dataset 1 (HDD), it takes 15 seconds. On huge dataset 3, it takes 35 minutes. Using a SSD leads to virtually no improvement.

Ideas for improving the implementation:

* Pre-load a big contiguous chunk of features_masks in memory (100,000 spikes or whatever) instead of fetching every spike straight from HDF5.
* Instead of preallocating the HDF5 cluster files, write the data to temporary flat binary files. Appending data to such files is much faster than appending to an HDF5 array. At the end, copy these binary files into the HDF5 cluster files.
* Write to the HDF5 cluster files in chunk instead of spike-per-spike.
* Forgetting about HDF5 completely and using flat binary files for the cluster store. However this is not trivial because we need to keep information about the array metadata somewhere, so we'll end up inventing our own file format.

opinions @cgestes @nippoo ?

**Comment by rossant** [comment]
Some more notes:

* More benchmarks make me thing that HDF5 is too damn slow for the cluster store. I think we need something else.
* One possibility would be to design our own file format, but that's risky.
* [bcolz](http://bcolz.blosc.org) is an interesting option. It can be easily installed with [conda](https://binstar.org/blaze/bcolz).
* Here is a [gist](https://gist.github.com/rossant/02dda655690ab3a2a92b) with a quick and dirty benchmark script.

@cgestes is that something you'd want to look into once you're done with the GUI? That would give me more time to make progress on the rest of the Python API.

Let me explain what to benchmark (see the gist above):

* Start from a big `(n_spikes, n_channels)` array in HDF5 (the KWX file, we can't change the format for now)
* You have a cluster assigned to each spike
* The end-result must be a set of files, one per cluster, containing the sub-array corresponding to the spikes in that cluster
* Eventually there will be several arrays in each file (typically 2, features and masks), but for benchmark it should be enough to use only one
* In the gist above, I process the data chunk by chunk, and I save the results incrementally on the disk (appending is fast on flat binary files)
* Finally, I put back the binary files into HDF5 (this is slow)

It would be better if we could get rid of the last step. So we want a unique file format such that:

* appending data is fast
* I/O is fast
* don't care about file size

For profiling you can use line_profiler. Don't forget to clear the OS cache too.

**Comment by rossant** [comment]
OK this is fixed in #160, closing.

The cluster store generation is now limited by I/O speed. It takes 3 minutes on Nick's dataset instead of more than 30 min with the previous implementation. Here is what I did:

* I switched from HDF5 to flat raw binary files for the cluster files (one file for the features, one for the masks)
* I don't store any metadata like dtype or shape: it is up to the code to know which shape and dtype to use. This makes everything much simpler.
* I go through the original features_masks array in chunks of 10,000-100,000 spikes.
